yarn-project/key-store/src/test_key_store.ts

```diff
@@ -233,21 +285,60 @@ export class TestKeyStore implements KeyStore {
    * @dev Used when feeding the master nullifier secret key to the kernel circuit for nullifier keys verification.
    */
   public getMasterNullifierSecretKeyForPublicKey(masterNullifierPublicKey: PublicKey): Promise<GrumpkinPrivateKey> {
-    // We iterate over the map keys to find the account address that corresponds to the provided public key
-    for (const [key, value] of this.#keys.entries()) {
-      if (value.equals(masterNullifierPublicKey.toBuffer()) && key.endsWith('-npk_m')) {
-        // We extract the account address from the map key
-        const accountAddress = key.split('-')[0];
-        // We fetch the secret key and return it
-        const masterNullifierSecretKeyBuffer = this.#keys.get(`${accountAddress.toString()}-nsk_m`);
-        if (!masterNullifierSecretKeyBuffer) {
-          throw new Error(`Could not find master nullifier secret key for account ${accountAddress.toString()}`);
-        }
-        return Promise.resolve(GrumpkinScalar.fromBuffer(masterNullifierSecretKeyBuffer));
+    // We get the account address associated with the master nullifier public key hash
+    const accountAddressBuffer = this.#keys.get(`${masterNullifierPublicKey.hash().toString()}-npk_m_hash`);
+    if (!accountAddressBuffer) {
+      throw new Error(
+        `Could not find account address for master nullifier public key ${masterNullifierPublicKey.toString()}`,
+      );
+    }
+    const accountAddress = AztecAddress.fromBuffer(accountAddressBuffer);
+
+    // We fetch the public keys and find this specific public key's position in the buffer
+    const masterNullifierPublicKeysBuffer = this.#keys.get(`${accountAddress.toString()}-np_keys_m`);
+    if (!masterNullifierPublicKeysBuffer) {
+      throw new Error(`Could not find master nullifier public keys for account ${accountAddress.toString()}`);
+    }
+
+    // We check that the buffer's length is a multiple of Point.SIZE_IN_BYTES
+    if (masterNullifierPublicKeysBuffer.byteLength % Point.SIZE_IN_BYTES !== 0) {
+      throw new Error("Master nullifier public key buffer's length is not a multiple of Point.SIZE_IN_BYTES.");
+    }
+
+    // Now we iterate over the public keys in the buffer to find the one that matches the hash
+    const numKeys = masterNullifierPublicKeysBuffer.byteLength / Point.SIZE_IN_BYTES;
```

Comment:
It seems like there is quite some overlap of the logic here, can likely be isolated to a small function to help clarity and reduce duplication.